src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseGitDiffReportService.java

```diff
@@ -0,0 +1,389 @@
+
+                        lastLineRemoveOperation = true;
+                        currentPreviousLineCount++;
+                    }
+                    case ' ' -> {
+                        currentEntry = handleUnchanged(entries, currentFilePath, currentPreviousFilePath, currentEntry);
+
+                        lastLineRemoveOperation = false;
+                        currentLineCount++;
+                        currentPreviousLineCount++;
+                    }
+                    default -> {
+                        deactivateCodeReading = true;
+                    }
+                }
+            }
+        }
+        if (!currentEntry.isEmpty()) {
+            entries.add(currentEntry);
+        }
+        return entries;
+    }
+
+    @NotNull
+    private ProgrammingExerciseGitDiffEntry handleUnchanged(List<ProgrammingExerciseGitDiffEntry> entries, String currentFilePath, String currentPreviousFilePath,
+            ProgrammingExerciseGitDiffEntry currentEntry) {
+        var entry = currentEntry;
+        if (!entry.isEmpty()) {
+            entries.add(entry);
+        }
+        entry = new ProgrammingExerciseGitDiffEntry();
+        entry.setFilePath(currentFilePath);
+        entry.setPreviousFilePath(currentPreviousFilePath);
+        return entry;
+    }
+
+    @NotNull
+    private ProgrammingExerciseGitDiffEntry handleRemoval(List<ProgrammingExerciseGitDiffEntry> entries, String currentFilePath, String currentPreviousFilePath,
+            ProgrammingExerciseGitDiffEntry currentEntry, boolean lastLineRemoveOperation, int currentPreviousLineCount) {
+        var entry = currentEntry;
+        if (!lastLineRemoveOperation && !entry.isEmpty()) {
+            entries.add(entry);
+            entry = new ProgrammingExerciseGitDiffEntry();
+            entry.setFilePath(currentFilePath);
+            entry.setPreviousFilePath(currentPreviousFilePath);
+        }
+        if (entry.getPreviousLineCount() == null) {
+            entry.setPreviousLineCount(0);
+            entry.setPreviousStartLine(currentPreviousLineCount);
+        }
+        entry.setPreviousLineCount(entry.getPreviousLineCount() + 1);
+        return entry;
+    }
+
+    private void handleAddition(ProgrammingExerciseGitDiffEntry currentEntry, int currentLineCount) {
+        if (currentEntry.getLineCount() == null) {
+            currentEntry.setLineCount(0);
+            currentEntry.setStartLine(currentLineCount);
+        }
+        currentEntry.setLineCount(currentEntry.getLineCount() + 1);
+    }
+
+    /**
+     * Extracts the file path from the raw git-diff for a specified diff block
+     *
+     * @param lines       All lines of the raw git-diff
+     * @param currentLine The line where the gitDiffLinePattern matched
+     * @return The file path of the current diff block
+     */
+    private String getFilePath(String[] lines, int currentLine) {
+        if (lines[currentLine - 1].startsWith("+++ ") && lines[currentLine - 2].startsWith("--- ")) {
```

Comment:
I suggest to add a check here, even if one thinks that it 'can't happen' :P
```suggestion
        if (currentLine > 1 && lines[currentLine - 1].startsWith("+++ ") && lines[currentLine - 2].startsWith("--- ")) {
```